File: leet/leet_581.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int findUnsortedSubarray(vector<int>& nums) {
        int max_int = ~(1 << 31);
        int min_int = (1 << 31);
        int len = nums.size();
        if(len == 1) return 0;
        if(len == 2) return (nums[0] > nums[1]) ? 2 : 0;
        int left_sorted_cnt = 0, right_sorted_cnt = 0;
        //int i = 0, j = len - 1 ;
        for(int i = 0; i < len - 1; ++i)
        {
            if(nums[i] <= nums[i+1]) 
            {
                ++left_sorted_cnt;
            }
            else break;
        }
        if(left_sorted_cnt == len - 1) return 0;
        for(int i = len - 1; i >= 1; --i)
        {
            if(nums[i] >= nums[i-1]) 
            {
                ++right_sorted_cnt;
            }
            else break;
        }
        // get min and max in the unsorted region
        int min_loc = max_int, max_loc = min_int;
        for(int i = left_sorted_cnt; i < len - right_sorted_cnt; ++i)
        {
            if(min_loc > nums[i]) min_loc = nums[i];
            if(max_loc < nums[i]) max_loc = nums[i];
        }
        // initial unsorted length
        int unsort_len = len - left_sorted_cnt - right_sorted_cnt;
        for(int i = left_sorted_cnt - 1; i >= 0; --i)
        {
            if(nums[i] > min_loc) ++unsort_len;
            else break;
        }
        for(int i = len - right_sorted_cnt; i < len; ++i)
        {
            if(nums[i] < max_loc) ++unsort_len;
            else break;
        }
        return unsort_len;
    }
};

int main()
{
    int a[] = {2, 6, 9, 15};
    vector<int> arr(a, a + sizeof(a)/sizeof(a[0]));
    Solution solu;
    cout << solu.findUnsortedSubarray(arr) << endl;
    return 0;
}

File: leet/leet_581.py
# find the shortest unsorted continuous subarray
class Solution(object):
    def findUnsortedSubarray(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        max_int = 2147483647
        min_int = -2147483648
        leng = len(nums)
        if leng == 1:
            return 0
        if leng == 2:
            return int(nums[0] > nums[1]) * 2 
        left = 0
        right = 0
        # check left border
        for i in range(leng - 1):
            if nums[i] <= nums[i+1]:
                left += 1
            else:
                break
        if left == leng - 1:
            return 0
        # check right border: from [leng-1] to [1]
        for i in range(leng - 1, 0, -1):
            if nums[i] >= nums[i-1]:
                right += 1
            else:
                break
        # get min and max in the unsorted region
        min_pos = max_int
        max_pos = min_int
        for i in range(left, leng - right, 1):
            if min_pos > nums[i]:
                min_pos = nums[i]
            if max_pos < nums[i]:
                max_pos = nums[i]
        # initial unsorted length
        unsort_len = leng - left - right
        for i in range(left - 1, -1, -1):
            if nums[i] > min_pos:
                unsort_len += 1
            else:
                break
        for i in range(leng - right, leng, 1):
            if nums[i] < max_pos:
                unsort_len += 1
            else:
                break
        return unsort_len

if __name__ == '__main__':
    solu = Solution()
    print solu.findUnsortedSubarray([2, 6, 5, 9, 15])
